test/performance/range/gap_decorator_rand_read_benchmark.cpp

```diff
@@ -0,0 +1,93 @@
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE
+// -----------------------------------------------------------------------------------------------------
+
+#include <chrono>
+#include <cmath>
+#include <cstring>
+#include <random>
+#include <utility>
+
+#include <benchmark/benchmark.h>
+
+#include <seqan3/alignment/aligned_sequence/aligned_sequence_concept.hpp>
+#include <seqan3/alignment/exception.hpp>
+#include <seqan3/alphabet/all.hpp>
+#include <seqan3/range/decorator/all.hpp>
+#include <seqan3/std/ranges>
+
+#include "gap_decorator_helper.hpp"
+
+using namespace seqan3;
+
+// Apply benchmarks with custom ranges for grid parameters sequence length and gap proportions
+static void CustomArguments(benchmark::internal::Benchmark* b) {
+    std::array<long long int, 5> gap_percentages = {1, 5, 25, 50, 75};
+    for (long long int seq_len = 4; seq_len <= (1 << 18); seq_len <<= 2)
+    {
+        for (auto gap_percentage : gap_percentages)
+            b->Args({seq_len, gap_percentage});
+    }
+}
+
+// ============================================================================
+//  read at random position
+// ============================================================================
+/* Parameters:
+ * gap_decorator_t      gap decorator class, e.g. gap_decorator_anchor_set
+ * gapped_flag          operate on already gapped (true) or ungapped sequence (false)
+ */
+template <typename gap_decorator_t, bool gapped_flag>
+static void read_random(benchmark::State& state)
+{
+    unsigned int seq_len = state.range(0);
+    using size_type = typename gap_decorator_t::size_type;
+    using sequence_type = remove_cvref_t<detail::unaligned_seq_t<gap_decorator_t>>;
+    sequence_type seq(seq_len, 'A'_dna4);
+
+    // vector of sampled gap lengths for each position
+    std::vector<size_type> gaps(seq_len, 0);
+
+    // determine sum of gaps and non-gap symbols for not exceeding targeted sequence length
+    if constexpr(gapped_flag)
+    {
+        sample<size_type>(gaps, seq_len, state.range(1)/100.0);
+        resize<size_type, sequence_type>(gaps, seq, seq_len);
+    }
+
+    // initialize with (truncated) sequence and insert gaps from left to right
+    gap_decorator_t gap_decorator;
+    assign_unaligned(gap_decorator, seq);
+
+    // insert gaps before starting benchmark
+    if constexpr(gapped_flag)
+        insert_gaps<gap_decorator_t>(gaps, gap_decorator, seq_len);
+
+    std::mt19937 generator(time(0)); //Standard mersenne_twister_engine seeded with current time
+    std::uniform_real_distribution<> uni_dis{0.0, static_cast<double>(seq_len)};
+
+    // sample read positions in advance
+    std::vector<size_t> access_positions;
+    access_positions.resize(1 << 18);
+    for (size_t i = 0; i < access_positions.size(); ++i)
+        access_positions[i] = uni_dis(generator);
```

Comment:
i would generate fewer positions and loop over them more often...

Reply:
> > With the varying gap densities (1, 5, 25, 50, 75 %) the benchmark is running for 40 minutes! Splitting sequential and random access functions won't be sufficient, esp. when the anchor_block approach will join the suite. Tested sequence lengths are 4, 16, 64, 256, 1024, 4096, 16384, 65536, and 262144 (= 1<<15). Any proposals where to reduce the test cases?
> 
>     * I would do a check for `SEQAN3_LONG_TESTS` and if that is defined, perform the full set of tests, otherwise only a subset. This macro could be added to the nightlies or just activated on-deman.
> 
>     * I would choose the subset as sequence lengths: 64, 256, 4096 and gap densities: 1, 5, 50; If this still takes to long, we can change these further.
> 
>     * To further reduce the run-time, I would also generate the test sequence/gap selection once for each parameter combination (but independent of the decorator/vector). This also improves comparability.

The first two points I implemented. The last one is tricky. You would typically use a fixture then, but the gap percentages and lengths are encoded in the state which is not available for the fixture class! Also, from my POV (or better state of knowledge) using a templated, fixture does not allow to have the test function externally defined (how it is currently). This allowed me to call it with various `gap_decorator versions. I could make it only running with defining the function body directly after the `BENCHMARK_TEMPLATE_DEFINE_F` macro like shown in google benchmark examples. However, this solution is inconvenient when testing with various types. Let's discuss this when you are back.

Reply:
Basically, I switched back to the previous version, where the data structures are initialized around the state loop. Template + CustomArguments + Test Class is not possible, because the definition needs to be inline (see answer here: https://stackoverflow.com/questions/56006025/how-to-use-a-templated-google-benchmark-fixture-with-custom-arguments-correctly/56010875?noredirect=1#comment98736674_56010875).